### DISCUSSION TITLE
Lexer design?

### DISCUSSION THREAD

**Discussion by prologic** [NicoNex/tau#22] [General]
Something i noticed while trying to figure out how to add basic support for [escape sequences](https://github.com/NicoNex/tau/pull/21) is the way the Lexer is designed, looks like you're using recursion?

For example:

https://github.com/NicoNex/tau/blob/6b87b8708a1eb3ede01e764d86b134c92ec47277/lexer/lexer.go#L142-L151

Is there a risk of blowing up the Go stack here or has Golang itself become quite elver here and uses tail recursion? 😅 (I haven't tested this lately!)

**Comment by prologic** [comment]
Actually re-reading this is _may_ not be recursion at all. I might be mistake, this looks like a state machine?

**Comment by NicoNex** [comment]
Yup, it's a state machine, it's very similar to the lexer that Go uses.

I really recommend watching the video [Lexical Scanning in Go](https://youtu.be/HxaD_trXwRE) by Rob Pike both to know more and because the video itself is awesome. 😄

**Comment by NicoNex** [comment]
But anyway for the escape sequences iirc there's a function in the Go standard library for that.

**Reply by prologic** [reply]
O'really? Do tell!

**Reply by NicoNex** [reply]
I searched and I couldn't find it, so I probably was wrong :/.

**Reply by prologic** [reply]
Can you take a look at my WIP branch and see if I'm going in the right direction fo rthis?

**Reply by NicoNex** [reply]
I think that another option that you could consider is to add the string parsing in `ast/string.go`.
In that way you don't need to touch the lexer at all and it should be way easier.
Also I was thinking of using `l.pos` of the lexer to produce better error messages in the future, so I wouldn't change the user input in the lexer itself.

**Reply by prologic** [reply]
Ahh that actually makes sense, I'll do that instead 👌
